### ISSUE TITLE
Comparison of payload hash for MC tag and online for all L1 tables

### ISSUE THREAD

**Issue by mulhearn** [cms-l1t-offline/cmssw#42]
@dburns7 has made a tool for top-level comparison of L1 payloads.

The tool simply reads all the L1 tables from the DB, then uses the common DB tool for dumping the has of each payload using this snippet:

   process.GlobalTag.DumpStat =  True

Now that the tool is working, we need to compare data tags to MC tags and follow up on all discrepancies.

The relevant deadline is the close of 76X ~ Oct 26.

Here is the overall status of Data tag vs MC tag comparisons:

Calo Trigger:
- L1CaloGeometryRecord (Probably Harmless but Requires Check)
- L1GctJetFinderParamsRcd (Harmless for Stage-1)
- L1HfRingEtScaleRcd (Wrong Online, Right in MC)
- L1HtMissScaleRcd (Wrong Online, Right in MC)
- L1RCTParametersRcd (Known Problem, confirmed fixed in 760_pre7)

GT Tables:
- L1GtParametersRcd (Vasile checking)
- L1GtPrescaleFactorsAlgoTrigRcd (MC always all 1)
- L1GtPrescaleFactorsTechTrigRcd (MC always all 1)
- L1GtPsbSetupRcd (Vasile checking)
- L1GtTriggerMaskAlgoTrigRcd (MC always all 0)
- L1GtTriggerMaskTechTrigRcd (MC always all 0)

Muon Tables:
- L1MuDTTFMasksRcd (MC all zero, Data has significant number masked, experts say shouldn't matter)
- L1MuGMTParametersRcd (hash different, content identical)
- L1RPCBxOrConfigRcd (hash different, content identical)
- L1RPCConeDefinitionRcd (complicated structure, need to add print function... assumed find)
- L1RPCConfigRcd (hash different, content identical)
- L1RPCHsbConfigRcd (hash different, content identical)
- L1MuCSCTFConfigurationRcd (hash different, content identical)

ALSO STILL TODO:  double check that no unexpected conditions over-riden in config files

**Comment by dburns7** [comment]
Missing headers (look up in cmslxr):
//#include "CondFormats/DataRecord/interface/L1JetEtScaleRcd.h"
//#include "CondFormats/L1TObjects/interface/L1JetEtScale.h"

//#include "CondFormats/DataRecord/interface/L1EmEtScaleRcd.h"
//#include "CondFormats/L1TObjects/interface/L1EmEtScale.h"

//#include "CondFormats/DataRecord/interface/L1HtMissScaleRcd.h"
//#include "CondFormats/L1TObjects/interface/L1HtMissScale.h"

//#include "CondFormats/DataRecord/interface/L1HfRingEtScaleRcd.h"
//#include "CondFormats/L1TObjects/interface/L1HfRingEtScale.h"

//#include "CondFormats/DataRecord/interface/L1GtTriggerMaskVetoAlgoTrigRcd.h"
//#include "CondFormats/L1TObjects/interface/L1GtTriggerMaskVetoAlgoTrig.h"

//#include "CondFormats/DataRecord/interface/L1CaloGeometryRcd.h"

//#include "CondFormats/DataRecord/interface/L1GtPrescaleFactorsAlgoTrigRcd.h"
//#include "CondFormats/L1TObjects/interface/L1GtPrescaleFactorsAlgoTrig.h"

//#include "CondFormats/DataRecord/interface/L1GtPrescaleFactorsTechTrigRcd.h"
//#include "CondFormats/L1TObjects/interface/L1GtPrescaleFactorsTechTrig.h"

//#include "CondFormats/DataRecord/interface/L1GtTriggerMaskAlgoTrigRcd.h"
//#include "CondFormats/L1TObjects/interface/L1GtTriggerMaskAlgoTrig.h"

//#include "CondFormats/DataRecord/interface/L1GtTriggerMaskTechTrigRcd.h"
//#include "CondFormats/L1TObjects/interface/L1GtTriggerMaskTechTrig.h"

//#include "CondFormats/DataRecord/interface/L1GtTriggerMaskVetoTechTrigRcd.h"
//#include "CondFormats/L1TObjects/interface/L1GtTriggerMaskVetoTechTrig.h"

//#include "CondFormats/DataRecord/interface/NumL1CondRcd.h"
//#include "CondFormats/L1TObjects/interface/NumL1Cond.h"

Not in O2O list:
CaloConfig.h
CaloParams.h
DTTFBitArray.h
GlobalStableParameters.h
L1CaloEtScale.h
L1GtAlgorithm.h
L1GtBoard.h
L1GtBptxTemplate.h
L1GtCaloTemplate.h
L1GtCastorTemplate.h
L1GtCondition.h
L1GtCorrelationTemplate
L1GtDefinitions
L1GtEnergySumTemplate
L1GtExternalTemplate
L1GtFwd
L1GtHfBitCountsTemplate
L1GtHfRingEtSumsTemplate
L1GtJetCountsTemplate
L1GtMuonTemplate
L1GtPrescaleFactors
L1GtPsbConfig
L1GtTriggerMask
L1GtTriggerMenuFwd
L1MuDTAssParam
L1MuDTEtaPattern
L1MuDTExtParam
L1MuPacking
L1MuScale
L1TriggerLutFile.h
LUT.h

**Comment by mulhearn** [comment]
Complete list of tables with different payloads for MC tags for online tags:

   < L1GctJetFinderParamsRcd : a1382a5ec3896a1c658f55326acd16719ee85a43
   ---
   > L1GctJetFinderParamsRcd : 1e406be10ad1c2604d5f20b85104ea2fe9976b67
   11c11 
   < L1GtParametersRcd : adeccfb5cef554aa74f9c4627eb05f5e7358a87d
   ---
   > L1GtParametersRcd : 358b7e94fb4dfe996946fa5add95f00a0fe22198
   13c13
   < L1GtPsbSetupRcd : bf31c439739a6b7274e449b87b669b4b86a7a74a
   ---
   > L1GtPsbSetupRcd : b659a0481299ec60f0ae3f190e1d2f15f36c6297
   17c17
   < L1GtTriggerMenuRcd : a9fd5e028ccd6da23a11afa97a8dfccae4d2dc2e
   ---
   > L1GtTriggerMenuRcd : e14a9b8cf676c67af246c441d51b56452743e619
   23c23
   < L1MuCSCTFConfigurationRcd : 1dde65b7136ca50878b219dc58bfaa768d9cded7
   ---
   > L1MuCSCTFConfigurationRcd : 69fca6f053bb7a62fcdef3d8c923082c30fb76b2
   35c35
   < L1MuDTTFMasksRcd : 029287adbf03569f4b0e1005962967d6bc993a19
   ---
   > L1MuDTTFMasksRcd : fbe21d76db9f95168d962d13682113499e8292ed
   41c41
   < L1MuGMTParametersRcd : 7399bfe52fc7db1fb2b914010b43dd66733cff2c
   ---
   > L1MuGMTParametersRcd : c67e43ed5ece955fe17c8bf2168e20b5fb8a66c2
   49c49
   < L1RCTChannelMaskRcd : 85e77710ebfc33d17c97f9559e4adad43dbea728
   ---
   > L1RCTChannelMaskRcd : e8877a780fa5a1d0b390a6b84566abd426bbe5f9
   53c53
   < L1RCTParametersRcd : c58cbeabad1424e30e54c88ce77ccd2bb2a0aa1e
   ---
   > L1RCTParametersRcd : 6be63d4dae0cbbca2de7600ba22b0f2a8ace7f91
   61c61
   < L1RPCHsbConfigRcd : e48260449d3135a62809c47469b8c94d6d8d3052
   ---
   > L1RPCHsbConfigRcd : 0d3157cffa67fd191d763b2cf2aef33938b0e297

**Comment by mulhearn** [comment]
@dburns7 remaining TODO on this:
1) use cmslxr to find the location of the objects not in CondFormats/L1TObjects. 
2) look at includes above and see how to print out the payloads above that are discrepant (e.g. L1MuGMTParametersRcd).
3) Extend (2) for all tables that have printout capability but make config driven, defaulted to off.

**Comment by mulhearn** [comment]
topic branch for this task is here:
https://github.com/cms-l1t-offline/cmssw/tree/l1t-configdump-CMSSW_7_6_0_pre5

**Comment by dburns7** [comment]
Update pushed.

**Comment by mulhearn** [comment]
Here is the latest comparison of data to MC tags, with Dustin's update to include more tables:

  < L1CaloGeometryRecord : 0691baf38bdaa9ac9adc4271e80ebcb08402483b
  > L1CaloGeometryRecord : 40c5191cdff0272e7c11d09db98338b154f65d4a6c6

  < L1GctJetFinderParamsRcd : 0fbb433d66b1e5c2755857fd5e071f2c83f84e92
  > L1GctJetFinderParamsRcd : a1382a5ec3896a1c658f55326acd16719ee85a438

  < L1GtParametersRcd : 358b7e94fb4dfe996946fa5add95f00a0fe22198
  < L1GtPrescaleFactorsAlgoTrigRcd : 635bf03c2b9091a308b324e3be26d5241898cb3c
  < L1GtPrescaleFactorsTechTrigRcd : ff4647fa76b39df81179ad5a90130a57bd810755
  < L1GtPsbSetupRcd : b659a0481299ec60f0ae3f190e1d2f15f36c6297
  > L1GtParametersRcd : adeccfb5cef554aa74f9c4627eb05f5e7358a87d
  > L1GtPrescaleFactorsAlgoTrigRcd : 0ed6a5ca575a55efced6f966589afa443b542e98
  > L1GtPrescaleFactorsTechTrigRcd : 03d5fdbf0adfb868e6f931ea151907787b62ffe9
  > L1GtPsbSetupRcd : bf31c439739a6b7274e449b87b669b4b86a7a74a

  < L1GtTriggerMaskAlgoTrigRcd : 8354a173264702ca9ad53df31dbbcb4dff7459d7
  < L1GtTriggerMaskTechTrigRcd : 37f8e148ddd893452d501892b52250bc6bd0f57c
  > L1GtTriggerMaskAlgoTrigRcd : dd3990f02eb99227c864da0b22e4d1dfafbe4327
  > L1GtTriggerMaskTechTrigRcd : cb669311be56423f3e244c064b8d37d281cea0af

  < L1GtTriggerMaskVetoTechTrigRcd : fe3e4f311dc654725e80dc74b1ee46266bab96e4
  < L1GtTriggerMenuRcd : 01b70e6f3cbb72a679aea6f8bbe69dffe567129e
  < L1HfRingEtScaleRcd : 5013103d90053732ce62bfae71cc75cda8f43a54
  < L1HtMissScaleRcd : 6ba9ced3eed48b446f619249a08c35bdff5f82f1
  > L1GtTriggerMaskVetoTechTrigRcd : cb669311be56423f3e244c064b8d37d281cea0af
  > L1GtTriggerMenuRcd : 8a6f3ff57fba07d222f0942c4f7428be6225b9ec
  > L1HfRingEtScaleRcd : d4d9d19322dfdda5acbf7c54ec3a8a3bfc5d1435
  > L1HtMissScaleRcd : 22175e1e6b187fc83d2dd047f2a163680637068f

  < L1MuDTTFMasksRcd : 0cb96011404091b85465d8c4a2e1637ed38ab741
  > L1MuDTTFMasksRcd : 029287adbf03569f4b0e1005962967d6bc993a19

  < L1MuGMTParametersRcd : 6b133f9e25eb0b9d963f5544af8c5dfcb82bcad4
  > L1MuGMTParametersRcd : 7399bfe52fc7db1fb2b914010b43dd66733cff2c

  < L1RCTParametersRcd : 6be63d4dae0cbbca2de7600ba22b0f2a8ace7f91
  < L1RPCBxOrConfigRcd : 87ee060ceeffb6dc4e26b6e4be1857b9ce362b95
  < L1RPCConeDefinitionRcd : f7ff667cde01f230a719932c5b6a6d3cd6af07fc
  < L1RPCConfigRcd : 1d1c8abc65a1ed26c3c616369a19500879b35a30
  < L1RPCHsbConfigRcd : a86ca0286c9677233cd63895dd054ffe87531574

  > L1RCTParametersRcd : c58cbeabad1424e30e54c88ce77ccd2bb2a0aa1e
  > L1RPCBxOrConfigRcd : 419271d4a6b37d08d04b4c0f98583716dd7eacee
  > L1RPCConeDefinitionRcd : f546f836ba316a399372561872879390dbed2da3
  > L1RPCConfigRcd : 1b2c234555c00affb378fa4f2082d5353cc25397
  > L1RPCHsbConfigRcd : e48260449d3135a62809c47469b8c94d6d8d3052

  < L1TriggerKeyListRcd : 2f62a302242d695d6b4f85f78f2713f9b9c4c359
  < L1TriggerKeyRcd  : 5d29478fb07f62fe6b051e965cd7096226e9926a
  > L1TriggerKeyListRcd : b0362017faeb5d0514b310cb805008b2046b2d51
  > L1TriggerKeyRcd : e05d7160fc71fcadee9f7ab80049dc9fb8a37398

**Comment by mulhearn** [comment]
@dburns7 time to practice your scripting: can you next write a perl or python script that parses the log file output from running your code and produces output like that above (Record Name : Hash Value).

**Comment by dburns7** [comment]
Done. Script attached and in /afs/cern.ch/user/d/dburns/work/public/for_mike

**Comment by mulhearn** [comment]
@dburns7 thanks for the script, will be testing shortly.

Now we need to urgently track down the discrepancies that your tool reveals.

**Comment by mulhearn** [comment]
We need by Monday to know if any conditions updates are needed for 76X, I'm reusing this branch to check.

**Comment by mulhearn** [comment]
@dburns7 you took me a bit too literally about reproducing the output ;-)  Can you make the script a bit more general and it outputs for any table with a hash code provided in the log file?

**Comment by mulhearn** [comment]
The discrepancy in JetFinder parameters should have no effect.  For Stage-1 we only use:

```
mulhearn@lxplus0020>grep HT DATA.LOG
GCT Jet Finder HT LSB GeV:  0.5
mulhearn@lxplus0020>grep HT MC.LOG
GCT Jet Finder HT LSB GeV:  0.5
```

**Comment by mulhearn** [comment]
For L1MuGMTParameters, I see different hashcodes:

```
L1MuGMTParametersRcd / : frontier://FrontierProd/CMS_CONDITIONS, L1MuGMTParameters_CRAFT09_hlt
  Time look up, payloadIds:
   255233 - 18446744073709551615 : 6b133f9e25eb0b9d963f5544af8c5dfcb82bcad4

L1MuGMTParametersRcd / : frontier://FrontierProd/CMS_CONDITIONS, L1MuGMTParameters_gmt2015_May27_min2BX_Latm2_DT9ne
    Time look up, payloadIds:
   1 - 18446744073709551615 : 7399bfe52fc7db1fb2b914010b43dd66733cff2c
```

But yet a 3-way agreement of actual parameter values between dump of data tag, dump of MC tag, and local config file...

```
EtaWeight_barrel:                0.028
PhiWeight_barrel:                 1
EtaPhiThreshold_barrel:           0.062
EtaWeight_endcap:                 0.13
PhiWeight_endcap:                 1
EtaPhiThreshold_endcap:           0.062
EtaWeight_COU:                    0.316
PhiWeight_COU:                    1
EtaPhiThreshold_COU:              0.127
CaloTrigger:                      1
IsolationCellSizeEta:     2
IsolationCellSizePhi:     2
DoOvlRpcAnd:                      0
PropagatePhi:             0
MergeMethodPhiBrl:                takeDT
MergeMethodPhiFwd:                takeCSC
MergeMethodEtaBrl:                Special
MergeMethodEtaFwd:                Special
MergeMethodPtBrl:                 byCombi
MergeMethodPtFwd:                 byCombi
MergeMethodChargeBrl:     takeDT
MergeMethodChargeFwd:     takeCSC
MergeMethodMIPBrl:                Special
MergeMethodMIPFwd:                Special
MergeMethodMIPSpecialUseANDBrl:  0
MergeMethodMIPSpecialUseANDFwd:  0
MergeMethodISOBrl:                Special
MergeMethodISOFwd:                Special
MergeMethodISOSpecialUseANDBrl:  1
MergeMethodISOSpecialUseANDFwd:  1
MergeMethodSRKBrl:                takeDT
MergeMethodSRKFwd:                takeCSC
HaloOverwritesMatchedBrl:         1
HaloOverwritesMatchedFwd:         1
SortRankOffsetBrl:                10
SortRankOffsetFwd:                10
CDLConfigWordDTCSC:               2
CDLConfigWordCSCDT:               3
CDLConfigWordbRPCCSC:     16
CDLConfigWordfRPCDT:              1
VersionSortRankEtaQLUT:           273
VersionLUTs:                      1
```

**Comment by mulhearn** [comment]
For L1MuDTTFMasksRcd, the MC masks are all zero, but the data masks are:

(Run 256650 to infty)

```
Disable CSC :
=============

 0 0 0 0 0 0 0 0 0 0 0 0
 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1
 0 0 0 0 0 0 0 0 0 0 0 0

Disable ETTF St.1 :
===================

 0 0 0 0 0 0 0 0 0 1 0 0
 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0

Disable ETTF St.2 :
===================

 0 0 0 0 0 0 1 0 0 0 0 0
 0 0 0 0 1 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0
 1 0 0 0 0 0 0 0 0 0 0 0
```

I guess the individual channel discrepancies are expected, but the CSC masks look significantly different...

**Comment by mulhearn** [comment]
For RPC configs, again despite the different hash codes, but data and MC dumps have the same values for every field:

```
L1RPCConfig: PPT:  12
L1RCPHsbConfig:
hsb0, hsb1
3, 3
3, 3
3, 3
3, 3
3, 3
3, 3
3, 3
3, 3
L1RPCBxOrConfig:
first bx:  0
last bx:   0
```

**Comment by mulhearn** [comment]
That only leaves RPCConeDefinition...  which has fairly complicated structure.

**Comment by mulhearn** [comment]
@apana can you add the snippet:

```
process.GlobalTag.DumpStat = True
```

and send me log file output when you run a few events through your configurations for Data and your configuration for MC?

**Comment by dburns7** [comment]
Update in log parser at /afs/cern.ch/user/d/dburns/work/public/for_mike

Dustin Burns
Physics Department
University of California, Davis

**Comment by mulhearn** [comment]
Update:  Discussing with Giacomo and Andreas, hash-code discrepancies can arise when payloads are written using different versions of Boost.  Confirmed for L1MuGMTParameters that the boost version was different for Data vs MC payloads.  So now it is understood how hash code can be different but payload is identical.

**Comment by kkotov** [comment]
![l1mugmtparametersrcd](https://cloud.githubusercontent.com/assets/11047399/10764026/940a0446-7ccc-11e5-9ac1-d2e776be92fc.png)
As Mike noticed, there may be a difference between 74x and 75x because the L1MuGMTParameters have been accepted for 75x, but not for the 74x.

**Comment by kkotov** [comment]
Well, a more detailed screenshot:
![l1tgmtparameters](https://cloud.githubusercontent.com/assets/11047399/10764162/28d31d42-7ccd-11e5-97e2-5a675c4e2283.png)

**Comment by mulhearn** [comment]
The latest list of discrepancies, comparing GTs 76X_mcRun2_asymptotic_v10 and 76X_dataRun2_v9
are: 

```
L1CaloGeometryRecord
L1GctJetFinderParamsRcd
L1GtParametersRcd
L1GtPrescaleFactorsAlgoTrigRcd
L1GtPrescaleFactorsTechTrigRcd
L1GtPsbSetupRcd
L1GtTriggerMaskAlgoTrigRcd
L1GtTriggerMaskTechTrigRcd
L1GtTriggerMaskVetoTechTrigRcd
L1GtTriggerMenuRcd
L1HfRingEtScaleRcd
L1HtMissScaleRcd
L1MuCSCTFConfigurationRcd
L1MuDTTFMasksRcd
L1MuGMTParametersRcd
L1RCTChannelMaskRcd
L1RPCBxOrConfigRcd
L1RPCConeDefinitionRcd
L1RPCConfigRcd
L1RPCHsbConfigRcd
```

Only new item on the list is L1MuCSCTFConfigurationRcd needs investigation.

**Comment by mulhearn** [comment]
Looks like CSC changed some configuration on 9-19-2015:

```
256839       2015-09-19 03:12:25  1dde65b7136ca50878b219dc58bfaa768d9cded7 L1MuCSCTFConfiguration  
256849       2015-09-19 16:00:03  ea13da4262accef13a63888e019447b14a20f31f  L1MuCSCTFConfiguration  
256858       2015-09-19 18:53:09  1dde65b7136ca50878b219dc58bfaa768d9cded7  L1MuCSCTFConfiguration  
256870       2015-09-20 02:28:48  ea13da4262accef13a63888e019447b14a20f31f  L1MuCSCTFConfiguration  
256914       2015-09-20 12:54:49  5350de365149d44824820423dd49ee0aabab7918  L1MuCSCTFConfiguration  
256927       2015-09-20 19:28:31  ea13da4262accef13a63888e019447b14a20f31f  L1MuCSCTFConfiguration  
256931       2015-09-20 23:56:17  5350de365149d44824820423dd49ee0aabab7918  L1MuCSCTFConfiguration  

```

**Comment by bortigno** [comment]
Hi Mike, In those dates CSCTF changed the timing of the CSC-DT trigger primitive exchange for a few sectors. to improve the synchronisation. There have been several configuration in a short time because we had to guess the direction of the corrections. We created different configuration for collision running and cosmic running - bottom only and all phi.

**Comment by mulhearn** [comment]
@bortigno:  will this effect the output of the emulation or is this not simulated?

**Comment by bortigno** [comment]
I am not sure timing is emulated. I think emulation assumes trigger primitives coming in time. A better synchronisation of real data would lead to a better agreement with the emulation. But I will double check on that.

**Comment by mulhearn** [comment]
All cleared up...  Here's what happened because its Edumacational.  Around 9-19 the CSC folks made some timing adjustments that had no effect on the offline table.  But they made new subsystem keys because, in fact, the online conditions did change.  So the O2O diligently did its job, constructing new payloads from the new keys.  Now, if we lived in a _sane_ world, at this point the hash code would have been calculated, it would be found to be identical to the previous version, and no new payload would be written, saving me a near heart attack one month later.  Instead, we live in an insane world, and the hashcodes the DB infrastructure calculates are Boost version dependent.  So identical payloads gave us different hashes.  I've confirmed explicitly that the payloads corresponding to hashes 5350de365149d44824820423dd49ee0aabab7918 and 1dde65b7136ca50878b219dc58bfaa768d9cded7 are identical.
